### PR TITLE
Support item count and operation name in azmonexporter

### PR DESCRIPTION
**Pull request by trask**
Introduces
* `AzureMonitorSpanProcessor` - handles operation name propagation
* `AzureMonitorSampler` - implements Azure Monitor trace-id hashing based sampler

Removes
* Agent's `AiSampler` which used tracestate to propagate sampling decision (and so required coordination with propagator to suppress propagating that tracestate since it wasn't a standard so didn't want to expose that publicly)

Fixes
* Agent's default sampling behavior to not be "parent-based" because that doesn't support accurate "item counts" since item count isn't propagated in tracestate (hopefully some day)
* Replaces configuration option `ignoreRemoteParentNotSampled` with better descriptive `parentBased` (under sampling preview)

Side effect:
* You can now specify higher sampling percentage to downstream spans, e.g. you can say that you want to capture 100% of a specific dependency, even if that is part of a trace that is not sampled-in. Can remove the warning note at https://docs.microsoft.com/en-us/azure/azure-monitor/app/java-standalone-sampling-overrides#how-it-works.

This behavior change around parent-based sampling probably deserves a minor version bump (to 3.4.0) and we should post follow-ups to #2152 and #1883 where the `ignoreRemoteParentNotSampled` was previously discussed.

### CODE REVIEW

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/AppIdSupplier.java [hunk 1]
```diff
@@ -53,8 +53,8 @@ public class AppIdSupplier implements AiAppId.Supplier {
       Executors.newSingleThreadScheduledExecutor(
           ThreadPoolUtils.createDaemonThreadFactory(AppIdSupplier.class));
 
-  private static final WarningLogger warningLogger =
-      new WarningLogger(GetAppIdTask.class, "Unable to retrieve appId");
+  private static final OperationLogger operationLogger =
```
Comment (heyams) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/AppIdSupplier.java:56]: 👍

File: agent/azure-monitor-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/AzureMonitorLogProcessor.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+import io.opentelemetry.api.common.AttributesBuilder;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.sdk.common.CompletableResultCode;
+import io.opentelemetry.sdk.logs.LogProcessor;
+import io.opentelemetry.sdk.logs.data.LogData;
+import io.opentelemetry.sdk.trace.ReadableSpan;
+
+public class AzureMonitorLogProcessor implements LogProcessor {
+
+  private static final AttributeKey<Long> AI_ITEM_COUNT_KEY =
```
Comment (heyams) [inline agent/azure-monitor-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/AzureMonitorLogProcessor.java:36]: this is already defined in SpanDataMapper.

File: agent/azure-monitor-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/AzureMonitorSpanProcessor.java [hunk 3]
```diff
@@ -19,25 +19,24 @@
...
 // name from routing instrumentation
 //       if we (only) set operation name on requests here, it would be based on span name at
 // startSpan
-public class AiOperationNameSpanProcessor implements SpanProcessor {
+public class AzureMonitorSpanProcessor implements SpanProcessor {
 
-  // visible for testing
-  static final AttributeKey<String> AI_OPERATION_NAME_KEY =
-      AttributeKey.stringKey("applicationinsights.internal.operation_name");
+  private static final AttributeKey<Long> AI_ITEM_COUNT_KEY =
```
Comment (heyams) [inline agent/azure-monitor-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/AzureMonitorSpanProcessor.java:38]: same here.. could reuse the one defined in SpanDataMapper.

File: agent/azure-monitor-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/LogDataMapper.java [hunk 4]
```diff
@@ -49,6 +48,8 @@ public class LogDataMapper {
 
   private static final AttributeKey<String> AI_OPERATION_NAME_KEY =
       AttributeKey.stringKey("applicationinsights.internal.operation_name");
+  private static final AttributeKey<Long> AI_ITEM_COUNT_KEY =
```
Comment (heyams) [inline agent/azure-monitor-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/LogDataMapper.java:51]: same here

File: agent/azure-monitor-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/LogDataMapper.java [hunk 5]
```diff
@@ -49,6 +48,8 @@ public class LogDataMapper {
 
   private static final AttributeKey<String> AI_OPERATION_NAME_KEY =
```
Comment (heyams) [inline agent/azure-monitor-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/LogDataMapper.java:49]: OperationNames.AI_OPERATION_NAME_KEY

File: agent/azure-monitor-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/OperationNames.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+ * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+package com.azure.monitor.opentelemetry.exporter.implementation;
+
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.sdk.trace.ReadableSpan;
+
+public class OperationNames {
```
Comment (heyams) [inline agent/azure-monitor-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/OperationNames.java:27]: should we make it final?

File: agent/azure-monitor-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/logging/OperationLogger.java [hunk 7]
```diff
@@ -63,7 +63,13 @@ public void recordFailure(String failureMessage, AzureMonitorMsgId msgId) {
   }
 
   // failureMessage should have low cardinality
-  public void recordFailure(String failureMessage, @Nullable Throwable exception) {
+  public void recordFailure(String failureMessage) {
+    if (aggregatingLogger != null) {
+      aggregatingLogger.recordWarning(failureMessage, null);
```
Comment (heyams) [inline agent/azure-monitor-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/logging/OperationLogger.java:68]:
```suggestion
      aggregatingLogger.recordFailure(failureMessage, null);
```

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/configuration/Configuration.java [hunk 8]
```diff
@@ -156,6 +156,28 @@ public static class Sampling {
...
+    //           will be wrong
+    //
+    // AND SO: if want to use parent-based sampler, then need to propagate the sample rate,
+    //         otherwise can end up with incorrect math
+    //
+    // Another (lesser) reason is because .NET SDK always propagates trace flags "00" (not
+    // sampled)
+    //
+    // IMPORTANT if changing this default, we need to keep it at least on Azure Functions
+    public boolean parentBased;
```
Comment (heyams) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/configuration/Configuration.java:179]: i don't find that this config is being check anywhere?

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/sampling/Samplers.java [hunk 9]
```diff
@@ -21,43 +21,28 @@
...
-        .build();
+  public static Sampler getSampler(float samplingPercentage, Configuration config) {
+    Sampler sampler = new AzureMonitorSampler(samplingPercentage);
+
+    if (!config.preview.sampling.overrides.isEmpty()) {
+      sampler =
+          new AiOverrideSampler(new SamplingOverrides(config.preview.sampling.overrides), sampler);
+    }
+
+    if (!config.preview.sampling.parentBased) {
```
Comment (heyams) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/sampling/Samplers.java:38]: got, it is checked here.

Comment (heyams) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/sampling/Samplers.java:38]: should this line before line 33? 

if it's parentBased, we don't care about sampling overrides.

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/sampling/Samplers.java [hunk 10]
```diff
@@ -21,43 +21,28 @@
...
-        // because this will lead to broken traces in A (sampled) -> B (not sampled) -> C (sampled)
-        // C will point to parent B, but B will not be exported
-        .setLocalParentNotSampled(Sampler.alwaysOff())
-        // can filter out subtree of sampled trace, by applying sampling override
-        .setRemoteParentSampled(parentSampledSampler)
-        // can filter out subtree of sampled trace, by applying sampling override
-        .setLocalParentSampled(parentSampledSampler)
-        .build();
+  public static Sampler getSampler(float samplingPercentage, Configuration config) {
+    Sampler sampler = new AzureMonitorSampler(samplingPercentage);
```
Comment (heyams) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/sampling/Samplers.java:31]: when parentBased is false, should the sampler be new `AzureMonitorSampler(samplingPercentage, false)`?
check line 38 before line 31?

File: agent/azure-monitor-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/AzureMonitorLogProcessor.java (hunk 2 above)
Comment (trask) [reply agent/azure-monitor-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/AzureMonitorLogProcessor.java]: went ahead and created `AiSemanticAttributes` class to consolidate these (will add more to that later)

File: agent/azure-monitor-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/logging/OperationLogger.java (hunk 7 above)
Comment (trask) [reply agent/azure-monitor-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/logging/OperationLogger.java:68]: there is no recordFailure(?)

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/sampling/Samplers.java (hunk 9 above)
Comment (trask) [reply agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/sampling/Samplers.java:38]: sampling overrides take precedence over parent-based, I added comment to the code

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/sampling/Samplers.java (hunk 10 above)
Comment (trask) [reply agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/sampling/Samplers.java:31]: I think the comment I added answers this, but let me know

File: agent/azure-monitor-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/logging/OperationLogger.java (hunk 7 above)
Comment (heyams) [reply agent/azure-monitor-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/logging/OperationLogger.java:68]: never mind.. it's in aggregatingLogger. Sorry, thought you overloaded the method below, line 72 for this.

**Review by heyams — APPROVED** [review]

**Pull request merged**